### ISSUE TITLE
Use a consistent naming convention for WA attributes specified on other elements

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#266]
**Current consensus:** Use a `data-wa-*` prefix (and let's try to avoid overly verbose names).

To-Do:
- [ ] Rename `data-toggle-nav` (possibly to `data-wa-toggle="nav"` to allow room for growth?)
- [ ] Rename `data-dialog`
- [ ] Convert popup's `data-current-placement` to a state (or two)

----

Even beyond https://github.com/shoelace-style/webawesome/discussions/202 WA *already* has a bunch of attributes it expects to be placed on other elements, such as:

- `data-toggle-nav` (part of `<wa-page>`)
- `data-dialog="close"` (part of `<wa-dialog>`)
- `data-current-placement` (part of `<wa-popup>`, output only so should probably be a `:state()`)
- whatever `disable-sticky` becomes, see #238  (part of `<wa-page>`)
- whatever `nav-state` becomes, see #238  (part of `<wa-page>`)
- which others?

How should these attributes be prefixed?
1. `data-`, like they are now
2. `wa-`
3. `data-wa-`
4. both 2 and 3

Pros & Cons:
- `data-` does not distinguish these attributes from any other attributes by other scripts or even the author's own attributes
- `data-wa-` is excessively long, but it may be okay as a secondary syntax if we go with `wa-` so that people don't complain about validation
- IMO `wa-` is the best: it communicates that these are parts of WA, it is 50% shorter than `data-`, and it is consistent with our other prefixes. Its only downside is that it doesn't validate (at least for now — there are many proposals to allow this pattern for custom attributes). There is plenty of precedent of JS frameworks and the like using custom prefixed and the vast majority of people don't care about validation.
- Supporting two naming schemes slightly complicates reading these, but not a huge deal
- Not using `data-` means we lose the `element.dataset` API, but not a huge deal.

So I think my top choice would be 4, followed by 2.

@claviska's [vote goes to 1, i.e. the current approach](https://github.com/shoelace-style/webawesome/issues/246#issuecomment-2514392655):

> data- prevents validation and type issues and it’s less to remember than invokers so I’m for it.

What do others think?

**Comment by claviska** [comment]
Note: we're only implementing `data-dialog="close"` right now. The `data-dialog="open <id>"` was a proposal to build on what we have, but not implemented. (It's trivial, however.)

**Comment by claviska** [comment]
Note that using `data-*` works around all possible validation/typing issues and is a valid use of the data attribute that, while not impossible, is unlikely to collide with other uses. Even `data-wa-dialog=""` is better despite being longer, IMO.

Ideally, I wish the platform would have settled on a better syntax for `command` and `commandfor`. We can still support those in the future no matter what we choose here, but having shorter sugar is indeed preferable.

```html
<wa-dialog>
  <wa-button data-dialog="close">Close</wa-button>
</wa-dialog>

<wa-dialog>
  <wa-button data-wa-dialog="close">Close</wa-button>
</wa-dialog>

<wa-dialog>
  <wa-button wa-dialog="close">Close</wa-button>
</wa-dialog>
```

While `wa-` is the shortest version, I worry that "wa-" is too strong an opinion and people will feel rather icky about it, even aside from validation/typing issues.

**Comment by lindsaym-fa** [comment]
It seems like we want to decide which of the following problems is the least painful for devs:
1. Lack of differentiation from other libraries or self-authored attributes (`data-`)
2. Lack of validation and native feature support (`wa-`)
3. Verbosity (`data-wa-`)
4. For teams especially, inconsistent usage (given devs have two options) and selective lack of validation (`wa-` and `data-wa-`)

To me, 1 seems like it might be the least painful, followed by 3, but I'd defer to all of you with more dev experience.

**Comment by LeaVerou** [comment]
That's an excellent summary of the tradeoffs @lindsaym-fa.

I started some polls because I genuinely have no idea what the right answer is:
- https://x.com/LeaVerou/status/1864380879327375397
- https://front-end.social/@leaverou/113596043240937128

It's still early, but so far `data-` is the _least_ preferred singular option in both and surprisingly, `data-wa-` the clear winner so far:
<img width="590" alt="image" src="https://github.com/user-attachments/assets/ccfc0078-e7f1-4152-9c48-eb5ac9eb5149">

<img width="399" alt="image" src="https://github.com/user-attachments/assets/b20ceb7b-f4ca-4fe2-99bb-1b4135c4c4ee">

I’ll update this post as more results come in.

**Comment by lindsaym-fa** [comment]
Personally, `data-wa-` would be my preference, too. I like the benefit of native support _and_ namespacing. I figured verbosity would be more of a turn off for devs, though!

**Comment by LeaVerou** [comment]
Notes from today's meeting:

- Need to investigate TypeScript consequences of `wa-*`
- Need to investigate what people *dislike* as well
	- https://x.com/LeaVerou/status/1864733650987995334
	- https://front-end.social/@leaverou/113601505435088783

**Comment by claviska** [comment]
Are we all settled on the `data-wa-*` pattern then? If so, I can update dialog/drawer.

To recap:

```html
<!-- Closes the containing dialog -->
<wa-button data-wa-dialog="close">

<!-- Opens the dialog with an id of "my-dialog" -->
<wa-button data-wa-dialog="open my-dialog">
```

**Comment by claviska** [comment]
Tracking in #246
